curp/src/sync_manager.rs

```diff
@@ -0,0 +1,166 @@
+    /// The endpoint to call rpc to other servers
+    ep: Endpoint,
+    /// Get cmd sync request from speculative command
+    sync_chan: MpscKeybasedReceiver<C::K, SyncMessage<C>>,
+    /// Send cmd to sync complete handler
+    comp_chan: SpmcKeybasedSender<C::K, SyncCompleteMessage>,
+    /// Other server address
+    others: Arc<RwLock<Vec<SocketAddr>>>,
+    /// Consensus log
+    log: Arc<Mutex<Vec<LogEntry<C>>>>,
+}
+
+impl<C: Command + 'static> SyncManager<C> {
+    /// Create a `SyncedManager`
+    pub(crate) fn new(
+        ep: Endpoint,
+        sync_chan: MpscKeybasedReceiver<C::K, SyncMessage<C>>,
+        comp_chan: SpmcKeybasedSender<C::K, SyncCompleteMessage>,
+        others: Arc<RwLock<Vec<SocketAddr>>>,
+        log: Arc<Mutex<Vec<LogEntry<C>>>>,
+    ) -> Self {
+        Self {
+            ep,
+            sync_chan,
+            comp_chan,
+            others,
+            log,
+        }
+    }
+
+    /// Run the `SyncManager`
+    pub(crate) async fn run(&mut self) {
+        let f = self.others.read().len().wrapping_div(2);
+
+        loop {
+            let sync_msg = match self.sync_chan.async_recv().await {
+                Ok(sync_msg) => sync_msg,
+                Err(_) => return,
+            };
+            let (term, cmd) = sync_msg.map_msg(SyncMessage::take_all);
+
+            let others: Vec<SocketAddr> = self
+                .others
+                .map_read(|others| others.iter().copied().collect());
+
+            let index = self.log.map_lock(|mut log| {
+                log.push(LogEntry::new(term, *(cmd.clone()), EntryStatus::Unsynced));
+                // length must be larger than 1
+                log.len().wrapping_sub(1)
+            });
+
+            let rpcs = others.iter().map(|addr| {
+                self.ep.call_timeout(
+                    *addr,
+                    SyncCommand::new(term, index.numeric_cast(), *(cmd.clone())),
+                    SYNC_TIMEOUT,
+                )
+            });
+
+            let mut rpcs: FuturesUnordered<_> = rpcs.collect();
+            let mut synced_cnt: usize = 0;
+
+            while let Some(resp) = rpcs.next().await {
+                let _result = resp
+                    .map_err(|err| {
+                        warn!("rpc error when sending `Sync` request, {err}");
+                    })
+                    .map(|r| {
+                        match r {
+                            SyncResponse::Synced => {
+                                synced_cnt = synced_cnt.overflow_add(1);
+                            }
+                            SyncResponse::WrongTerm(_)
+                            | SyncResponse::EntryNotEmpty(_)
+                            | SyncResponse::PrevNotReady(_) => {
+                                // todo
+                            }
+                        }
+                    });
+
+                if synced_cnt == f {
+                    let _r = self
```

Comment:
Do we need to check Result?

Reply:
Yes, this version channel returns error, we should abort sync manager and unrecoverable.